`NullPointerException` when retrieving `@ManyToOne` composite table with additional properties in `@IdClass`

When retrieving an entity that models a composite table between two other tables and uses an `@IdClass` that includes the IDs of the related entities **and an additional property** as the composite primary key, the query fails with a `NullPointerException`. The issue occurs in version `>=2.4.0.Final`, but not in `2.3.1.Final`.

For example, given the following simple entities:
```java
@Entity
@Table(name = "product")
public class Product {

    @Id
    @Column(name = "id")
    private UUID id;

    @Column(name = "name")
    private String name;

    // Constructor, Getter and Setter
}
```

```java
@Entity
@Table(name = "customer")
public class Customer {

    @Id
    @Column(name = "id")
    private UUID id;

    @Column(name = "name")
    private String name;

    // Constructor, Getter and Setter
}
```

The entities are connected through the following relationship entity, which uses an `@IdClass`:
```java
@Entity
@Table(name = "delivery")
@IdClass(DeliveryId.class)
public class Delivery {

    @Id
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", referencedColumnName = "id")
    private Product product;

    @Id
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id", referencedColumnName = "id")
    private Customer customer;

    @Id
    @Column(name = "additional_key")
    private String additionalKey;

    // Constructor, Getter, Setter, Equals, HashCode
}
```

```java
public class DeliveryId implements Serializable {

    private UUID product;
    private UUID customer;
    private String additionalKey;

   // Getter, Setter, Equals, HashCode
}
```

When persisting exemplary entities and retrieving the `Delivery` relationship with a query such as
```java
session.createQuery("SELECT d FROM Delivery d", Delivery.class).getResultList()
```

the following exception is thrown:
```
12:01:25.612 [vert.x-eventloop-thread-16] INFO  o.h.r.v.impl.DefaultVertxInstance - HR000003: Vert.x instance stopped
Exception in thread "main" java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "org.hibernate.reactive.sql.results.graph.ReactiveInitializer.getData(org.hibernate.sql.results.jdbc.spi.RowProcessingState)" because "initializer" is null
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1177)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2341)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:216)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:411)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:382)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:175)
        at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveNonAggregatedIdentifierMappingInitializer.reactiveResolveKey(ReactiveNonAggregatedIdentifierMappingInitializer.java:64)
        at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveNonAggregatedIdentifierMappingInitializer.reactiveResolveKey(ReactiveNonAggregatedIdentifierMappingInitializer.java:25)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.initializeId(ReactiveEntityInitializerImpl.java:749)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:625)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:605)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:56)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.resolveKey(ReactiveStandardRowReader.java:383)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.lambda$coordinateInitializers$28(ReactiveStandardRowReader.java:376)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$7(CompletionStages.java:410)
        at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:484)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:121)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:102)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:197)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:411)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:382)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.coordinateInitializers(ReactiveStandardRowReader.java:376)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.reactiveReadRow(ReactiveStandardRowReader.java:126)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$addToResultsSupplier$5(ReactiveListResultsConsumer.java:141)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$consume$1(ReactiveListResultsConsumer.java:96)
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at io.vertx.core.Future.lambda$toCompletionStage$3(Future.java:601)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:88)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:32)
        at io.vertx.core.Promise.complete(Promise.java:66)
        at io.vertx.core.Promise.handle(Promise.java:51)
        at io.vertx.core.Promise.handle(Promise.java:29)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:60)
        at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:298)
        at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:169)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:57)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
        at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
        at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:324)
        at io.vertx.pgclient.impl.PgSocketConnection.handleMessage(PgSocketConnection.java:114)
        at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:137)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:328)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:321)
        at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:388)
        at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        at io.vertx.pgclient.impl.codec.PgDecoder.fireCommandResponse(PgDecoder.java:52)
        at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:133)
        at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:248)
        at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:107)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "org.hibernate.reactive.sql.results.graph.ReactiveInitializer.getData(org.hibernate.sql.results.jdbc.spi.RowProcessingState)" because "initializer" is null
        at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveNonAggregatedIdentifierMappingInitializer.lambda$reactiveResolveKey$1(ReactiveNonAggregatedIdentifierMappingInitializer.java:69)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$2(CompletionStages.java:175)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$7(CompletionStages.java:410)
        at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:484)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:121)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:102)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:197)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:215)
        ... 79 more
```

Please find the following reproducer, which can be executed with `./gradlew run`: [hibernate-reactive-idclass-issue.zip](https://github.com/user-attachments/files/17291581/hibernate-reactive-idclass-issue.zip)

Best regards,
Janek

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveNonAggregatedIdentifierMappingInitializer.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveNonAggregatedIdentifierMappingInitializer.java
@@ -62,7 +62,7 @@ public class ReactiveNonAggregatedIdentifierMappingInitializer extends NonAggreg
 			final RowProcessingState rowProcessingState = data.getRowProcessingState();
 			final boolean[] dataIsMissing = {false};
 			return loop( getInitializers(), initializer -> {
-				if ( dataIsMissing[0] ) {
+				if ( dataIsMissing[0] || initializer == null ) {
 					return voidFuture();
 				}
 				final InitializerData subData = ( (ReactiveInitializer<?>) initializer )

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveNonAggregatedIdentifierMappingInitializer.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveNonAggregatedIdentifierMappingInitializer.java
@@ -60,9 +60,8 @@ public class ReactiveNonAggregatedIdentifierMappingInitializer extends NonAggreg
 		}
 		else {
 			final RowProcessingState rowProcessingState = data.getRowProcessingState();
-			final boolean[] dataIsMissing = {false};
 			return loop( getInitializers(), initializer -> {
-				if ( dataIsMissing[0] || initializer == null ) {
+				if ( initializer == null ) {
 					return voidFuture();
 				}
 				final InitializerData subData = ( (ReactiveInitializer<?>) initializer )
@@ -72,7 +71,6 @@ public class ReactiveNonAggregatedIdentifierMappingInitializer extends NonAggreg
 						.thenAccept( v -> {
 							if ( subData.getState() == State.MISSING ) {
 								data.setState( State.MISSING );
-								dataIsMissing[0] = true;
 							}
 						} );
 			} );

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneIdClassTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneIdClassTest.java
@@ -1,0 +1,141 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import java.util.Collection;
+import java.util.List;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.vertx.junit5.Timeout;
+import io.vertx.junit5.VertxTestContext;
+import jakarta.persistence.Entity;
+import jakarta.persistence.FetchType;
+import jakarta.persistence.Id;
+import jakarta.persistence.IdClass;
+import jakarta.persistence.ManyToOne;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Timeout(value = 1, timeUnit = MINUTES)
+public class ManyToOneIdClassTest extends BaseReactiveTest {
+
+	private final static String USER_NAME = "user";
+	private final static String SUBSYSTEM_ID = "1";
+
+	@Override
+	protected Collection<Class<?>> annotatedEntities() {
+		return List.of( SystemUser.class, Subsystem.class );
+	}
+
+	@BeforeEach
+	public void populateDb(VertxTestContext context) {
+		Subsystem subsystem = new Subsystem( SUBSYSTEM_ID, "sub 1" );
+		SystemUser systemUser = new SystemUser( subsystem, USER_NAME, "system 1" );
+		test(
+				context, getMutinySessionFactory()
+						.withTransaction( (s, t) -> s.persistAll( subsystem, systemUser ) )
+		);
+	}
+
+	@AfterEach
+	public void after(VertxTestContext context) {
+		test( context, cleanDb() );
+	}
+
+	@Test
+	public void testQuery(VertxTestContext context) {
+		test(
+				context, openSession()
+						.thenAccept( session -> session.createQuery( "SELECT s FROM SystemUser s", SystemUser.class )
+								.getResultList().thenAccept( list -> {
+									assertThat( list.size() ).isEqualTo( 1 );
+									SystemUser systemUser = list.get( 0 );
+									assertThat( systemUser.getSubsystem().getId() ).isEqualTo( SUBSYSTEM_ID );
+									assertThat( systemUser.getUsername() ).isEqualTo( USER_NAME );
+								} ) )
+		);
+	}
+
+	@Entity(name = "SystemUser")
+	@IdClass(PK.class)
+	public static class SystemUser {
+
+		@Id
+		@ManyToOne(fetch = FetchType.LAZY)
+		private Subsystem subsystem;
+
+		@Id
+		private String username;
+
+		private String name;
+
+		public SystemUser() {
+		}
+
+		public SystemUser(Subsystem subsystem, String username, String name) {
+			this.subsystem = subsystem;
+			this.username = username;
+			this.name = name;
+		}
+
+		public Subsystem getSubsystem() {
+			return subsystem;
+		}
+
+		public String getUsername() {
+			return username;
+		}
+
+		public String getName() {
+			return name;
+		}
+	}
+
+	@Entity(name = "Subsystem")
+	public static class Subsystem {
+
+		@Id
+		private String id;
+
+		private String description;
+
+		public Subsystem() {
+		}
+
+		public Subsystem(String id, String description) {
+			this.id = id;
+			this.description = description;
+		}
+
+		public String getId() {
+			return id;
+		}
+
+		public String getDescription() {
+			return description;
+		}
+	}
+
+	public static class PK {
+
+		private Subsystem subsystem;
+
+		private String username;
+
+		public PK(Subsystem subsystem, String username) {
+			this.subsystem = subsystem;
+			this.username = username;
+		}
+
+		private PK() {
+		}
+	}
+
+}

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneIdClassTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneIdClassTest.java
@@ -8,7 +8,6 @@ package org.hibernate.reactive;
 import java.util.Collection;
 import java.util.List;
 
-import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
@@ -40,26 +39,21 @@ public class ManyToOneIdClassTest extends BaseReactiveTest {
 		SystemUser systemUser = new SystemUser( subsystem, USER_NAME, "system 1" );
 		test(
 				context, getMutinySessionFactory()
-						.withTransaction( (s, t) -> s.persistAll( subsystem, systemUser ) )
+						.withTransaction( s -> s.persistAll( subsystem, systemUser ) )
 		);
-	}
-
-	@AfterEach
-	public void after(VertxTestContext context) {
-		test( context, cleanDb() );
 	}
 
 	@Test
 	public void testQuery(VertxTestContext context) {
 		test(
-				context, openSession()
-						.thenAccept( session -> session.createQuery( "SELECT s FROM SystemUser s", SystemUser.class )
-								.getResultList().thenAccept( list -> {
-									assertThat( list.size() ).isEqualTo( 1 );
-									SystemUser systemUser = list.get( 0 );
-									assertThat( systemUser.getSubsystem().getId() ).isEqualTo( SUBSYSTEM_ID );
-									assertThat( systemUser.getUsername() ).isEqualTo( USER_NAME );
-								} ) )
+				context, openSession().thenAccept( session -> session
+						.createQuery( "FROM SystemUser s", SystemUser.class )
+						.getResultList().thenAccept( list -> {
+							assertThat( list ).hasSize( 1 );
+							SystemUser systemUser = list.get( 0 );
+							assertThat( systemUser.getSubsystem().getId() ).isEqualTo( SUBSYSTEM_ID );
+							assertThat( systemUser.getUsername() ).isEqualTo( USER_NAME );
+						} ) )
 		);
 	}
 

